hail/python/hail/backend/service_backend.py

```diff
@@ -537,127 +479,43 @@ def _cancel_on_ctrl_c(self, coro: Awaitable[T]) -> T:
                 self._batch_was_submitted = False
             raise
 
-    def execute(self, ir: BaseIR, timed: bool = False, **kwargs):
-        return self._cancel_on_ctrl_c(self._async_execute(ir, timed=timed, **kwargs))
-
-    async def _async_execute(self,
-                             ir: BaseIR,
-                             *,
-                             timed: bool = False,
-                             progress: Optional[BatchProgressBar] = None,
-                             **kwargs):
-        async def inputs(infile, token):
-            await write_int(infile, ServiceBackend.EXECUTE)
-            await write_str(infile, tmp_dir())
-            await write_str(infile, self.billing_project)
-            await write_str(infile, self.remote_tmpdir)
-            await write_str(infile, self.render(ir))
-            await write_str(infile, token)
-            await write_int(infile, len(self.functions))
-            for fun in self.functions:
-                await fun.serialize(infile)
-            await write_str(infile, '{"name":"StreamBufferSpec"}')
-
-        resp, timings = await self._rpc(
-            'execute(...)',
-            inputs,
-            ir=ir,
-            progress=progress,
-            **kwargs
+    def _rpc(self, action: ActionTag, payload: ActionPayload) -> Tuple[bytes, str]:
+        return self._cancel_on_ctrl_c(self._async_rpc(action, payload))
+
+    async def _async_rpc(self, action: ActionTag, payload: ActionPayload):
+        if action == ActionTag.EXECUTE:
+            assert isinstance(payload, ExecutePayload)
```

Comment:
This kind of pattern makes me wonder if the tag should just hang off of the payload e.g. `.tag`?

Reply:
Ya I kind of agree, except on the server I switch on the tag to determine how I should deserialize the payload, so it didn't make sense to me for that to be part of the payload.

Reply:
Hmm. I leave your comment up to you. I'm OK with this all as it is and don't have my head deep enough in this to have strong opinions about whether ir functions and tokens are in execute or outside.

Reply:
I think I'll leave as is now and we can revisit in a separate PR